include/envoy/event/dispatcher.h

```diff
@@ -115,6 +116,12 @@ class Dispatcher {
    */
   virtual Event::TimerPtr createTimer(TimerCb cb) PURE;
 
+  /**
+   * Create a thread.
+   * @param thread_routine supplies the function to invoke in the thread.
+   */
+  virtual Thread::ThreadPtr createThread(std::function<void()> thread_routine) PURE;
```

Comment:
I'm not sure how threads relate to dispatchers. Can you talk a little more about what you are trying to do here?

Reply:
@eziskind can probably explain better, but I think the idea is provide a way to do dependency injection for thread creation; at Google we have a different threading library.

There is a chicken-and-egg thing with dispatchers and threading; threads run dispatchers. That said, where would be a better place to inject?

Reply:
We (Google) would like to add some additional instrumentation when creating threads. The idea is to allow injecting a thread factory class via the MainCommonBase class, with a default implementation which just does a "new ThreadImpl" which is what happens today. Wiring through the dispatcher seems like the least disruptive way to allow accessing the thread factory class (when it gets added in a future change).

Reply:
I see. Yeah the issue is that I don't think it's guaranteed that we have a dispatcher in all places we create a thread (maybe we do in all the places you care about). FWIW my original intention was to do this kind of stuff in https://github.com/envoyproxy/envoy/blob/master/include/envoy/api/api.h, and that would be the way we could also had the rest of the system from extensions. So my feeling would be to add this as part of API and then inject there? Obviously, then you might need API in a bunch of places, which leaves us with either exposing API out of dispatcher (maybe less ugly) or doing some time of singleton thing like we do for OsSysCalls. No great solution here. Thoughts?

Reply:
The Watchdog is an example of something that uses a thread not connected to a dispatcher: https://github.com/envoyproxy/envoy/blob/master/source/server/guarddog_impl.h